Add subscan operations to preprocess

earosenberg opened the pull request:
Add basic statistics, PSDs, and glitches on subscans.

msilvafe commented on sotodlib/tod_ops/flags.py:
```diff
@@ -672,3 +692,151 @@ def get_inv_var_flags(aman, signal_name='signal', nsigma=5,
+    return arr[ndslice]
+
+def wrap_info(aman, info_aman_name, info, info_names, merge=True):
+    """Wrap info into an aman. Assumed to be (dets,) or (dets, subscans)"""
```
This function at first glance seems to be quite general but I think it's somewhat specific to wrapping in the stats info appropriately choosing whether or not to include subscan axes. Can you just call it `wrap_stats` or `wrap_and_check_subscans` or something similarly specific and give the doc string a bit more description.

msilvafe commented on sotodlib/tod_ops/flags.py:
```diff
@@ -672,3 +692,151 @@ def get_inv_var_flags(aman, signal_name='signal', nsigma=5,
+    stat_names : list
+        List of strings identifying which statistics to run.
+    split_subscans : bool
+        (Optional). If True statistics will be computed on subscans. Assumes aman.subscans exists already.
```
`aman.subscan_info`

msilvafe commented on sotodlib/tod_ops/flags.py:
```diff
@@ -672,3 +692,151 @@ def get_inv_var_flags(aman, signal_name='signal', nsigma=5,
+        signal = aman[signal]
+    if split_subscans:
+        if mask is not None:
+            raise ValueError("Cannot mask samples and split subscans")
```
It should be possible to do a union of the subscan flags + a secondary flag if they're both ranges() aligned along samps. Ok to leave for later but I'd just create an issue for it to track it.

earosenberg replied:
Added support for discontinuous ranges to `apply_rng` but didn't implement the full masking. Will add an issue

msilvafe reviewed (changes requested):
Thanks @earosenberg, it'll be really nice to have this infrastructure for subscan operations. I put inline comments reflecting the conversation we had on Friday plus a few more things as I went through more closely.

earosenberg commented on sotodlib/tod_ops/fft_ops.py:
```diff
@@ -281,6 +280,44 @@ def calc_psd(
+            freqs, pxx_sub = welch(signal_ss, fs, **kwargs)
+            Pxx.append(pxx_sub)
+        else:
+            Pxx.append(np.full((signal.shape[0], kwargs["nperseg"]//2+1), np.nan)) # Add nans if subscan is too short
```
I wanted Pxx to be `(nusamps_ss, subscans)`; same logic in the nan padding of `get_stats`. If a particular subscan is deemed too short or has been completely cut, due to restricts on `samps`, then nans get put in instead. I suppose we could instead do a restrict on the subscan axis, or I'm open to other suggestions.

msilvafe replied:
Ok, I think this is sensible.

earosenberg commented on sotodlib/tod_ops/fft_ops.py:
```diff
@@ -353,6 +390,8 @@ def fit_noise_model(
     merge_name="noise_fit_stats",
     merge_psd=True,
+    freq_spacing=None,
+    approx_fit=False,
```
Sounds good.
It's a little confusing that `calc_wn` returns the square root of `median(psd)` whereas `fit_noise_model` doesn't take the root; do we want to change that? Also seems cleaner to add a new function `estimate_fknee` but I can mash it all into `calc_wn` if you prefer

msilvafe replied:
Yes, this bothers me quite a lot that we (probably I) decided to change the convention between `calc_wn` and `fit_noise_model`. I think that they should both be in squared units (like `fit_noise_model` is now). I see what you're saying about wanting to have another function like `estimate_fknee` but I think my preference is to have one function called `estimate_noise_model` which returns exactly the same shaped output as `fit_noise_model` i.e. an axis shaped (dets x noise_coeffs). If we want to keep `calc_wn` fn alive for backwards compatibility we could just have it call `estimate_noise_model` (so that we don't have to update definitions of median/mean/etc in multiple places) and just return the white noise coefficients but I would keep everything in the squared units for consistency.

Another related issue was brought up by @ykyohei in https://github.com/simonsobs/sotodlib/issues/995 where we really should be always taking the PSDs w/ overlap = 0, storing the nperseg in the calc_psd output so that we can have the calc_wn/any fn that takes a median over psd bins correct for this "median bias". 

It's maybe ok to address that all here but also ok opening up a new PR for these fixes and reviewing that separately, I'll leave it up to you.

earosenberg replied:
After some consideration I think `estimate_noise_model` either needs to be more robust then just the initial guess of the `fit_noise_model` or at least have some error quantification. Otherwise you can have catastrophic mis-fits  -- this is also a problem with `fit_noise_model` but I assume it could be caught with the covariance. I've just deleted `approx_fit` for now, will open a new PR (or at least an issue) to deal with these other problems.

earosenberg commented:
Addressed all above points except where noted. Also added new features:
- `CalcPsd` with `subscan=True` now creates `psd.Pxx_ss` with a subscans axis and `psd.Pxx` from the mean of the subsans.
- Added plots for signal and psd

msilvafe reviewed (approved):
Thanks for addressing all of my comments and testing all of this.
